primitives/npos-elections/src/balancing.rs

```diff
@@ -0,0 +1,202 @@
+	if elected_edges.len() <= 1 {
+		return Zero::zero()
+	}
+
+	// amount of stake from this voter that is used in edges.
+	let stake_used = elected_edges
+		.iter()
+		.fold(0, |a: ExtendedBalance, e| a.saturating_add(e.borrow().weight));
+
+	// backed stake of each of the elected edges.
+	let backed_stakes = elected_edges
+		.iter()
+		.map(|e| e.borrow().candidate.borrow().backed_stake)
+		.collect::<Vec<_>>();
+
+	// backed stake of all the edges for whom we've spent some stake.
+	let backing_backed_stake = elected_edges
+		.iter()
+		.filter_map(|e|
+			if e.borrow().weight > 0 {
+				Some(e.borrow().candidate.borrow().backed_stake)
+			} else {
+				None
+			}
+		)
+		.collect::<Vec<_>>();
+
+	let difference = if backing_backed_stake.len() > 0 {
+		let max_stake = backing_backed_stake
+			.iter()
+			.max()
+			.expect("vector with positive length will have a max; qed");
+		let min_stake = backed_stakes
+			.iter()
+			.min()
+			.expect("iterator with positive length will have a min; qed");
+		let mut difference = max_stake.saturating_sub(*min_stake);
+		difference = difference.saturating_add(voter.budget.saturating_sub(stake_used));
+		if difference < tolerance {
+			return difference;
+		}
+		difference
+	} else {
+		voter.budget
+	};
+
+	// remove all backings.
+	for edge in elected_edges.iter_mut() {
+		{
+			let edge = edge.borrow();
+			let mut candidate = edge.candidate.borrow_mut();
+			candidate.backed_stake = candidate.backed_stake.saturating_sub(edge.weight);
+		}
+		{
+			edge.borrow_mut().weight = 0;
+		}
+	}
+
+	elected_edges.sort_unstable_by_key(|e| e.borrow().candidate.borrow().backed_stake);
+
+	let mut cumulative_backed_stake = Zero::zero();
+	let mut last_index = elected_edges.len() - 1;
+
+	for (index, edge) in elected_edges.iter().enumerate() {
+		let index = index as ExtendedBalance;
+		let backed_stake = edge.borrow().candidate.borrow().backed_stake;
+		let temp = backed_stake.saturating_mul(index);
+		if temp.saturating_sub(cumulative_backed_stake) > voter.budget {
+			// defensive only. length of elected_edges is checked to be above 1.
+			last_index = index.saturating_sub(1) as usize;
+			break
+		}
+		cumulative_backed_stake = cumulative_backed_stake.saturating_add(backed_stake);
+	}
+
+	let last_stake = elected_edges.get(last_index).expect(
+		"length of elected_edges is greater than or equal 2; last_index index is at \
+		the minimum elected_edges.len() - 1; index is within range; qed"
+	).borrow().candidate.borrow().backed_stake;
+	let ways_to_split = last_index + 1;
+	let excess = voter.budget
+		.saturating_add(cumulative_backed_stake)
+		.saturating_sub(last_stake.saturating_mul(ways_to_split as ExtendedBalance));
+
+	// Do the final update.
+	for edge in elected_edges.into_iter().take(ways_to_split) {
+		// first, do one scoped borrow to get the previous candidate stake.
+		let candidate_backed_stake = {
+			let edge_ref = edge.clone();
+			let edge = edge_ref.borrow();
```

Comment:
why do you clone here ?
```suggestion
			let edge = edge.borrow();
```